### DISCUSSION TITLE
行过滤和列脱敏

### DISCUSSION THREAD

**Discussion by yiluoya** [apache/doris#7539] [Ideas]
column masking
example：

customer table as follow ，the telephone num is sensitive ，don't want be select by user A。 but user B can view all data。

![![![image](https://user-images.githubusercontent.com/90463865/147719094-9e2c8049-04c7-4ba6-9394-32dfb8b811a1.png)](https://user-images.githubusercontent.com/90463865/147719026-0d430da0-dbed-4ce7-8852-52d9a58c209c.png)](https://user-images.githubusercontent.com/90463865/147718995-fc598f9c-1285-4de4-83f0-6e3942e8643a.png)
solution ：
add function at table customer column telephone for user A。
then user A do ： select * from customer :
![image](https://user-images.githubusercontent.com/90463865/147719001-e56f3421-17e0-4150-8ed8-7a39d1d60601.png)

at user B's view： select * from customer :
![image](https://user-images.githubusercontent.com/90463865/147718995-fc598f9c-1285-4de4-83f0-6e3942e8643a.png)

row filter

example ：table order
![image](https://user-images.githubusercontent.com/90463865/147719026-0d430da0-dbed-4ce7-8852-52d9a58c209c.png)

hope user A only can view data ‘省份’ is ‘湖南’。

solution ：
add express at table order for user A
then when user A do ： select * from order
![image](https://user-images.githubusercontent.com/90463865/147719039-f48791a2-b507-4ad5-92b8-cf33800619d2.png)

next is i implement example：

```
`CREATE database ods;
use ods;

CREATE TABLE test_chen
(
a varchar(10)
,b int
)
ENGINE=olap
DISTRIBUTED BY HASH(a) BUCKETS 32
PROPERTIES(
"storage_type"="column",
"replication_num"="1"
);

insert into test_chen values("1",1);
insert into test_chen values("2",2);
insert into test_chen values("3",-1);`
```

row filter ：
```
>> GRANT FILTER_PRIV ON ods.test_chen TO 'root' WITH a > 1;

SELECT * FROM ods.test_chen tc ;
```

![image](https://user-images.githubusercontent.com/90463865/147719065-1ac6c990-0a7f-45a5-abb5-cee830795e75.png)

column mask：

> GRANT MASK_PRIV ON ods.test_chen TO 'root' COLUMN b WITH abs(b);

![image](https://user-images.githubusercontent.com/90463865/147719083-d8871d72-1c02-4f78-9d09-a8606a01376f.png)
show grants

> SHOW GRANTS for 'root'

![image](https://user-images.githubusercontent.com/90463865/147719072-c11a5628-1ee9-4c16-a429-438eaf2fe361.png)
priv revoke

> REVOKE FILTER_PRIV ON ods.test_chen FROM 'root' REVOKE MASK_PRIV ON ods.test_chen FROM 'root' COLUMN b

![image](https://user-images.githubusercontent.com/90463865/147719088-db5df85c-f24c-4f4d-9302-e3540038e022.png)
propose:
is it a good way use mask_priv/filter_priv to grant ?

for huawei gsDB

https://support.huaweicloud.com/devg2-dws/dws_0402_0042.html

> SELECT DBMS_REDACT.add_policy

for hive /presto user range
![image](https://user-images.githubusercontent.com/90463865/147719175-b64e0b3d-6896-4cd4-9c09-2ece01bbb120.png)

**Comment by ChPi** [comment]
问下你们脱敏是放在了be层吗？
我们之前也做过权限和脱敏但都是上层处理

**Comment by yiluoya** [comment]
sql 重写，在fe层

**Reply by abcchen2** [reply]
请问重写是在SQL语句解析前还是解析后？
